src/datasources/db/postgres-database.migration.hook.ts

```diff
@@ -0,0 +1,25 @@
+import { Inject, Injectable, OnModuleInit } from '@nestjs/common';
+import postgres from 'postgres';
+import { ILoggingService, LoggingService } from '@/logging/logging.interface';
+import * as shift from 'postgres-shift';
+
+/**
+ * The {@link PostgresDatabaseMigrationHook} is a Module Init hook meaning
+ * that it will be executed once the dependencies are resolved.
+ *
+ * This happens before the Application Bootstraps, so route listeners are not
+ * initiated and potentially generating queries.
+ */
+@Injectable({})
+export class PostgresDatabaseMigrationHook implements OnModuleInit {
+  constructor(
+    @Inject('DB_INSTANCE') private readonly db: postgres.Sql,
+    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+  ) {}
+
+  async onModuleInit() {
+    this.loggingService.info('Checking migrations');
+    await shift({ sql: this.db });
```

Comment:
> This is probably out of the scope of the PR, but have you thought about what would be the proper behavior if an error happens while running the migrations (inside `shift`)? Maybe an error at this level should prevent the service from starting.

Promise should be rejected but I will test this condition. Good point 👍 

> Also an additional concern regarding this (sorry for the delay between comments): should we prevent the case where several instances of the service execute the migrations when starting? (is there a database lock or other mechanism that prevents concurrent execution?)

The migrations are executed inside a transaction but after looking more into it I think we should have a lock acquired on the application level. Advisory locks seem to be a good fit for this – https://postgresql.org/docs/16/interactive/explicit-locking.html#ADVISORY-LOCKS.

Reply:
I've implemented the lock here: [7fbe2cb](https://github.com/safe-global/safe-client-gateway/pull/829/commits/7fbe2cb56845e474800d93d9952e58c4ebf11e23)

Application also fails to start if the migration is not executed correctly.

One issue though is that if we have two instances and the following happens:
1. Instance 1 acquires lock
2. Instance 2 fails to acquire lock and returns
3. Instance 1 fails migrations and exits
4. Instance 2 is accepting connections

A possible solution would be to have instances always wait on the lock and perform the migration. This should be ok since migrations should check if a given table or operation was already executed. Given a single database instance the end result should be the same if the migrations are idempotent – either all instances fail or all are successful.

What do you think?

Reply:
Good point. I guess we can assume the majority of migrations will be idempotent, we could make that a requirement for a migration. If we need to include a non-idempotent migration in the future for some reason, we would need to consider a different way of persisting its changes to the database. 

Regarding the scenario you described, I'd say we need to test carefully in dev environments what happens when a bad migration is executed and causes the service to stop the bootstrap sequence. I mean, thinking in Kubernetes terms, I guess the failing instance 1 gets restarted repeatedly, but instance 2 as you said accepts connections. I think this scenario is acceptable because at least we don't have a service disruption. But it's also true instance 2 executes code that expects the migration to have been applied.

It's a complex problem, so as I said maybe we need to actually test these error conditions in a dev environment to have a better picture of the implications of waiting/exiting/erroring, etc.

Reply:
Right now I'm more leaning towards every instance tries to perform the pending migrations. Since they should be idempotent, after the migration is executed we should see no changes from the other executions and they should terminate successfully.

In this case I would say that it's more on the deployment side that at least one instance should remain alive (previous service version) if the migrations were not completed successfully.

This implies changing from `pg_try_advisory_lock` (no lock wait) to `pg_advisory_lock` (lock wait).

Reply:
> In this case I would say that it's more on the deployment side that at least one instance should remain alive (previous service version) if the migrations were not completed successfully.

Agree 100%. We need to check this is the case (i.e. the service never responds OK to the readiness checks if the migrations fail).

Reply:
I've changed it in [877e31f](https://github.com/safe-global/safe-client-gateway/pull/829/commits/877e31fafa70ccbb0b08685188a446522a0782af)